### ISSUE TITLE
Double clicking error in generated code in error list doesn't navigate to generated source

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/roslyn#69014]
This looks like a regression in Preview 3.

If a source-generated code contains an error, double clicking the error in Error List used to navigate to the generated code, but in Preview 3 it doesn't.

**Comment by Youssef1313** [comment]
I'm still very impacted by this 😕

**Comment by channell** [comment]
This is a pathetic bug introduced in a preview when the path for generated content was moved from within the source tree to a temp path, but diagnostic messages were not.  It is trivial to fix, and insulting to the community to ignore.
The only justifications are either weight of more important bugs, SourceGeneration being deprecated, or expectation that we fix it ourselves and  submit a pull request.

**Comment by CyrusNajmabadi** [comment]
> The only justifications are either weight of more important bugs

That is currently the case.  However, we'd love to work with you if this is impacting you heavily.  We def take and appreciate community contributions.  Thanks! :)

> we fix it ourselves and submit a pull request

Community fixes are a great way to unblock important issues you're running into that are currently lower priority than the current set of work on our plates.  If interested, let us know.  Tnx :)

**Comment by channell** [comment]
Source Generators used to be a great feature of domain-specific-languages that are translated to C# for inclusion in assemblies.  If a DSL need to validate that all generated code will compile without issue, it might as well generate the assembly directly.  In my case, all classes are generated as partial classes (to use C# functions for calculations) so can not do this - a simpler alternative is to do what Source Generated used to do, and add a MSBuild task to generate source to a ../gen directory (less effort than digging into the Rosyln code).

When I have time, I'll create a small reproducible case using Razor pages, and submit a bug-report through the license support channel

**Comment by CyrusNajmabadi** [comment]
>  a simpler alternative is to do what Source Generated used to do, and add a MSBuild task to generate source to a ../gen directory (less effort than digging into the Rosyln code).

That would be my *strong* recommendation.  SGs are for a very unique and specialized scenario, and are not intended as a general purpose solution for any sort of code gen task.  If your scenario doesn't need that specific functionality, and can work with an MSBuild task, you're likely better off using that.

Thanks!

**Comment by CyrusNajmabadi** [comment]
> When I have time, I'll create a small reproducible case using Razor pages, and submit a bug-report through the license support channel

Please just submit here.  We'll have to go find any other reports and route them here anyways, so it will save some annoying middle steps (which often can disconnect notification messages from the original filer).

**Comment by channell** [comment]
.

> > a simpler alternative is to do what Source Generated used to do, and add a MSBuild task to generate source to a ../gen directory (less effort than digging into the Rosyln code).
> 
> That would be my _strong_ recommendation. SGs are for a very unique and specialized scenario, and are not intended as a general purpose solution for any sort of code gen task. If your scenario doesn't need that specific functionality, and can work with an MSBuild task, you're likely better off using that.
> 
> Thanks!

If you have a reference for "SGs are for a very unique and specialized scenario, and are not intended as a general purpose solution" I would be interested to see it, otherwise please refrain from hallucination

**Comment by CyrusNajmabadi** [comment]
> > If you have a reference for "SGs are for a very unique and specialized scenario

https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview

> A Source Generator is a new kind of component that C# developers can write that lets you do two major things:
>
> 1. Retrieve a compilation object that represents all user code that is being compiled. This object can be inspected, and you can write code that works with the syntax and semantic models for the code being compiled, just like with analyzers today.
> 
> 1. Generate C# source files that can be added to a compilation object during compilation. In other words, you can provide additional source code as input to a compilation while the code is being compiled.

Please remember to follow the .net code of conduct: https://dotnetfoundation.org/about/policies/code-of-conduct.  Thanks :)

**Comment by channell** [comment]
> > > If you have a reference for "SGs are for a very unique and specialized scenario
> 
> https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview
> 
> > A Source Generator is a new kind of component that C# developers can write that lets you do two major things:
> > 
> > 1. Retrieve a compilation object that represents all user code that is being compiled. This object can be inspected, and you can write code that works with the syntax and semantic models for the code being compiled, just like with analyzers today.
> > 2. Generate C# source files that can be added to a compilation object during compilation. In other words, you can provide additional source code as input to a compilation while the code is being compiled.
> > 
> > ![image](https://private-user-images.githubusercontent.com/4564579/339920641-3d7f246d-b71c-493f-b64a-bf625e3a5e70.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg0NTE1MDAsIm5iZiI6MTcxODQ1MTIwMCwicGF0aCI6Ii80NTY0NTc5LzMzOTkyMDY0MS0zZDdmMjQ2ZC1iNzFjLTQ5M2YtYjY0YS1iZjYyNWUzYTVlNzAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTVUMTEzMzIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTdkMDUyMGEwZWYzMjU1MGMxZTE3YWIxNzgwNjhkODNhNmM0NmUyMGI4YWNmNDlkZDAyNDkwNWQ0MTk2ODczNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.FNXa2PBmt-7wBEFl5XDAKwgeRE9oW10kPGmJzddU35c)
> 
> As per above, SGs are for a very unique and specialized scenario, and are not intended as a general purpose solution for any sort of code gen task. They are intended only when this approach is the only suitable way to get generated sources to be effectively compiled into a program. There are many other approaches that are much simpler, and suited to many scenarios that don't have this special requirement.
> 
> Please remember to follow the .net code of conduct: https://dotnetfoundation.org/about/policies/code-of-conduct. Thanks :)

Source Generators is an Innovative solution to problems like {Razor pages, XAML based UI code, native/COM interop, serialisation code} that could to be generated at compile time, when historically dynamic code generation would have been used. Being compiled into the same assembly as related code means that internal classes can be referenced, and serialisation does not need to use reflection

The concept is closely related to Type Providers in F# and Cargo build in Rust, but has additional constraints that developers must respect the environment: IO (unlike F#) is forbidden, exceptions should not be thrown,  excessive CPU or memory utilisation should be avoided as this Interface loads executable code into the compilation service host. 

It is worth noting that generated code should not be considered "user-readable", and #line directives should be used to redirect users back to source code they can change.  My use of code-generators is exactly the same as System.Text.Json serialisation and Razor pages, and follows the use-case highlighted in your quote: It is a Hallucination to believe "very unique and specialized scenario" or that a problem arises because of inappropriate usage rather than an accidental bug introduced whilst addressing an unrelated a CI issue.  

It is worth  commenting that non-microsoft source generators can not be supported by Microsoft, and may need to be removed when reporting issues with the Roslyn compiler, that the change was introduced because it is poor practice to generate source code into the directory provided by GIT code checkout, and source navigation is only available in Visual Studio.

At no point was your "advice" requested, I simply want to know if the issue is going to be addressed, and if so, how quickly (my issue was closed as a "duplicate").  It is worth noting that the .net code of conduct: https://dotnetfoundation.org/about/policies/code-of-conduct applies to everyone, "hallucination" was used in the modern context of Large-Language-Models inferring something that does not exist rather than substance influence, and apologise if that was your impression.

**Comment by channell** [comment]
The actual solution is to use [EmitCompilerGeneratedFiles](https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration-generator) in the PropertyGroup section of the csproj file. Adding this .NET 6 feature resolved the issue - it is a shame that the maintenance engineers didn't know the product well enough before posting "working as designed" comments in error

Adding 
```
  <PropertyGroup>
    <EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>
  </PropertyGroup>
```
resolves the problem, with Visual studio using `/obj/[Target]/[Platform]/generated/[add-in]/[SourceGenerator]` as the default path for dotnet error messages.  There remains a minor bug that omission of `EmitCompilerGeneratedFiles` results in invalid error messages, but on balance it is probably not worth fixing.

Please close the issue as **resolved**

**Comment by CyrusNajmabadi** [comment]
@channell this doesn't resolve this issue.  The feature should work, even without that item in the propertygroup.

**Comment by channell** [comment]
The remaining issue is that the path in the error message is not the same as the path to the file as explained in [#73935](https://github.com/dotnet/roslyn/issues/73935)

> [severity:It's more difficult to complete my work] [regression] [worked-in:Noticed in 17.9.6, was not in 17.9.3]
When using Roslyn Code analyzers, an Error in the generated code is reported as
1>C:\dev\source\repos\HiperspaceHub\examples\Plan\Plan\HiLang\HiLang.Roslyn.SourceGenerator\Timesheet.g.cs(506,30,506,36): error CS0115: 'ValueType.Unbind(SubSpace)': no suitable method found to override
but the actual file is
C:\Users\steve\AppData\Local\Temp\VSGeneratedDocuments\3d2af891-ec5d-ccc2-8856-70b3b266fe5d\Timesheet.g.cs

`EmitCompilerGeneratedFiles` corrects the problem because the actual path is printed in the error message and double-click then works.  
It is not a regression, it's a "feature" because the file (`3d2af891-ec5d-ccc2-8856-70b3b266fe5d\Timesheet.g.cs` in my example) is a temporary file that will be deleted, possibly before the error is inspected - it can't be fixed (but a better error message would help), hence the suggestion to close the issue

**Comment by jasonmalinowski** [comment]
@channell Are you clicking the error list or the output window? The error list at least we're supposed to see it's a generated file and still open it in the dynamic view we create in that case.

**Comment by CyrusNajmabadi** [comment]
> EmitCompilerGeneratedFiles corrects the problem

That's a workaround.  The core feature is still something that would be worth fixing. 

> it can't be fixed 

It can be fixed.  

>hence the suggestion to close the issue

We should keep this open as long as there's a poor experience here.  I'm glad you have a mitigation.  But being able to click on error messages (even if to temp locations due to in-memory source gen files) is something that makes sense to work @channell

**Comment by Youssef1313** [comment]
I agree with @CyrusNajmabadi 

> It can be fixed.

Not only "can" be fixed, but I'm pretty sure it even **actually worked** at some point and regressed during (I think) .NET 8 previews.

**Comment by channell** [comment]
Why have I commented? Because  the Visual Studio paid subscription bug report was moved to [73935](https://github.com/dotnet/roslyn/issues/73935), and closed as a duplicate of this issue:
When was the defect introduced? After VS 17.9.3.

Why was it introduced? most likely due to a build server issue with the original location of generated files
Why won't it be fixed? 
* because the temp directory is temp and may be removed, so can't be used for error messages
* Defaulting to ./obj will render the MSBuild setting `EmitCompilerGeneratedFiles` redundant.
* Errors in generated files need to be fixed in the source generator anyway

It's a trivial change, that would be quicker to fix by forking the roslyn compiler, and a lot less effort that arguing about.

**Comment by Youssef1313** [comment]
> that would be quicker to fix by forking the roslyn compiler

While I'm not part of the Roslyn team, I'm sure they will happily review a reasonable community contribution to fix this. So if you're willing to look into it and fix it, you can open a Pull Request. The contribution will be very much appreciated.

**Comment by CyrusNajmabadi** [comment]
> Why won't it be fixed?

I would take a reasonable fix for this issue.  

>because the temp directory is temp and may be removed, so can't be used for error messages

This isn't relevant.  Neither are the other two points.

>  and a lot less effort that arguing about.

There's no reason to argue.  We're an open source project @channell .  We take community contributions all the time.  Otherwise, we get through issues based on priority and as the backlog allows.
